### PR TITLE
Package Integrity Checks 2

### PR DESCRIPTION
**Pull request by TingALin**
Add workspace structure integrity (pre&post)checks to main.yml: 

Pre-build checks:
 Every package/package.json name field should match polkadotcloud-<package folder name>.
 Every package/package.json scripts field should have build:mock, build and clear properties.

Post-build checks:
 Every package should have a package.json file, with @polkadotcloud/<package folder name> as the name field.

Addresses https://github.com/paritytech/polkadot-cloud/issues/343

### CODE REVIEW

**Review by wirednkod — COMMENTED** [review]
Some small extra checks are needed IMO

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -25,7 +25,9 @@ jobs:
         with:
           node-version: ${{ matrix.node-version }}
       - run: yarn
+      - run: yarn prebuild
       - run: yarn build
+      - run: yarn postbuild
```
Comment (rossbulat) [inline .github/workflows/main.yml:30]: You don't need to add `prebuild` and `postbuild`, they are called automatically upon running `build`.
```suggestion
      - run: yarn build
```

File: scripts/postbuild.mjs [hunk 2]
```diff
@@ -4,6 +4,30 @@
 import fs from "fs";
 import { dirFilesExist } from "./index.mjs";
 
+const matchName = async (dir, files) => {
+  for (let file of files) {
+    fs.stat(`${dir}${file}/dist/package.json`, (err) => {
+      if (err) {
+        console.error(`❌ dist/package.json directory not found`);
```
Comment (rossbulat) [inline scripts/postbuild.mjs:11]:
```suggestion
        console.error(`❌ dist/package.json file not found`);
```

File: scripts/postbuild.mjs [hunk 3]
```diff
@@ -4,6 +4,30 @@
...
+        return;
+      }
+      const json = JSON.parse(
+        fs.readFileSync(`${dir}${file}/dist/package.json`).toString()
+      );
+
+      const name = Object.entries(json).filter((p) => {
+        return "name".includes(p[0]);
+      });
```
Comment (rossbulat) [inline scripts/postbuild.mjs:20]:
```suggestion
      const name = Object.keys(json).filter((k) => 
         "name".includes(k)
      );
```

File: scripts/postbuild.mjs [hunk 4]
```diff
@@ -4,6 +4,30 @@
...
+
+      if (name[0][1] != `@polkadotcloud/${json.name.split(/\/(.*)/s)[1]}`) {
+        console.error(
+          `❌ ${name[0][1]} dist/package name doesn't meet the naming requirement`
```
Comment (rossbulat) [inline scripts/postbuild.mjs:24]:
```suggestion
          `❌ ${name[0][1]} Could not find `name` field in provided package.json`
```

File: scripts/postbuild.mjs [hunk 5]
```diff
@@ -4,6 +4,30 @@
...
+      }
+      const json = JSON.parse(
+        fs.readFileSync(`${dir}${file}/dist/package.json`).toString()
+      );
+
+      const name = Object.entries(json).filter((p) => {
+        return "name".includes(p[0]);
+      });
+
+      if (name[0][1] != `@polkadotcloud/${json.name.split(/\/(.*)/s)[1]}`) {
```
Comment (rossbulat) [inline scripts/postbuild.mjs:22]:
```suggestion
      if (json.name !== `@polkadotcloud/${json.name.split(/\/(.*)/s)[1]}`) {
```

Comment (rossbulat) [inline scripts/postbuild.mjs:22]:
```suggestion
      if (json?.name !== `@polkadotcloud/${json.name.split(/\/(.*)/s)[1]}`) {
```

File: scripts/postbuild.mjs (hunk 3 above)
Comment (rossbulat) [inline scripts/postbuild.mjs:20]:
```suggestion
```

File: scripts/postbuild.mjs (hunk 4 above)
Comment (rossbulat) [inline scripts/postbuild.mjs:24]:
```suggestion
          `❌ package.json `name` field does not match the naming requirement`
```

File: scripts/prebuild.mjs [hunk 6]
```diff
@@ -18,6 +18,42 @@ const dirFoldersOnly = async (dir, files) => {
   }
 };
 
+const matchNameNScripts = async (dir, files) => {
```
Comment (rossbulat) [inline scripts/prebuild.mjs:21]:
```suggestion
const matchNameNScripts = (dir, files) => {
```

File: scripts/prebuild.mjs [hunk 7]
```diff
@@ -35,6 +71,7 @@ try {
       "package.json",
       "lib",
     ]);
+    await matchNameNScripts("./packages/", files);
```
Comment (rossbulat) [inline scripts/prebuild.mjs:74]:
```suggestion
   matchNameNScripts("./packages/", files);
```

**Comment by TingALin** [comment]
> Some small extra checks are needed IMO

I have pushed the updated fix, please help to review it again :)

File: scripts/postbuild.mjs [hunk 8]
```diff
@@ -4,6 +4,26 @@
...
+    fs.stat(`${dir}${file}/dist/package.json`, (err) => {
+      if (err) {
+        console.error(`❌ dist/package.json file not found in ${dir}${file}`);
+        return;
+      }
+      const json = JSON.parse(
+        fs.readFileSync(`${dir}${file}/dist/package.json`).toString()
+      );
+
+      if (json?.name != `@polkadotcloud/${file}`) {
```
Comment (rossbulat) [inline scripts/postbuild.mjs:18]:
```suggestion
      if (json?.name !== `@polkadotcloud/${file}`) {
```

File: scripts/postbuild.mjs [hunk 9]
```diff
@@ -4,6 +4,26 @@
...
+        console.error(`❌ dist/package.json file not found in ${dir}${file}`);
+        return;
+      }
+      const json = JSON.parse(
+        fs.readFileSync(`${dir}${file}/dist/package.json`).toString()
+      );
+
+      if (json?.name !== `@polkadotcloud/${file}`) {
+        console.error(
+          `❌ package.json in ${json?.name}'s name field does not match the naming requirement`
```
Comment (rossbulat) [inline scripts/postbuild.mjs:20]:
```suggestion
          `❌ package.json name field does not match the naming requirement`
```

File: scripts/prebuild.mjs [hunk 10]
```diff
@@ -18,6 +18,34 @@ const dirFoldersOnly = async (dir, files) => {
   }
 };
 
+const matchNameNScripts = (dir, files) => {
```
Comment (rossbulat) [inline scripts/prebuild.mjs:21]: This name is a bit off, `matchScripts` is probably all we need.

File: scripts/prebuild.mjs [hunk 11]
```diff
@@ -18,6 +18,34 @@ const dirFoldersOnly = async (dir, files) => {
...
+      );
+
+      if (json?.name !== `polkadotcloud-${file}`) {
+        console.error(
+          `❌ ${json?.name} package name doesn't meet the naming requirement`
+        );
+      }
+
+      if (
+        !Object.keys(json?.scripts).includes("build:mock" && "build" && "clear")
```
Comment (rossbulat) [inline scripts/prebuild.mjs:39]: I don't think this `includes` is being used correctly, the inner definition will result to `true`? use `indexOf` instead.

**Comment by rossbulat** [comment]
@TingALin this now needs to be amended because the actual `npm` packages do not have the `cloud-` prefix in the name. You can use the same code to test the name:

```
  let publishedName = json.name.split(/-(.*)/s)[1];
  if (publishedName.startsWith("cloud-")) {
    // remove `cloud-` from the start of publishedName.
    publishedName = publishedName.slice("cloud-".length);
  }
```

**Review by rossbulat — APPROVED** [review]

**Pull request merged**